### ISSUE TITLE
New JS interop doesn't include prototype fields (Dartium only)

### ISSUE THREAD

**Issue by johnpryan** [dart-lang/sdk#24737]
D3.js returns a Selection object when using methods like `select()`.  A Selection object is an Array with additional fields defined on it's prototype

Running [this example](https://github.com/johnpryan/d3dart/blob/master/example/dart_example.dart) in Dartium results in an exception (it works in Chrome):

```
Exception: NoSuchMethodError: method not found: 'append'
Receiver: Instance of 'JsArrayImpl'
Arguments: ["svg"]
```

It looks like the prototype methods are only available on the `JavaScript View`:

<img width="449" alt="screen shot 2015-10-26 at 9 35 18 pm" src="https://cloud.githubusercontent.com/assets/1145719/10748041/02eaa3aa-7c2a-11e5-98fb-92703053094e.png">

**Comment by jacob314** [comment]
Thanks for trying out dart-js interop.
The good news is your example is just one line from working correctly in Dartium as well as dart2js.
You just need to specify that Selection implements List and then Dartium will allow you to call the Selection methods of Array objects.
We need to document this requirement.  Generally if you have classes that are Arrays you want to indicate they extend List so that your Dart IDE can show that they have all the List methods etc. 
I've changed the milestone to  1.14 as we need to document this behavior more or perhaps relax the constraints so that you aren't required to add the "implements List" in Dartium. It is something that was added in Dartium purely to try to make bugs where someone forgot to add "implements List" easier to catch. However, the downside is we can't easily make this case fail in dart2js.

https://codereview.appspot.com/269670043
show how to fix your example.

**Comment by johnpryan** [comment]
Thanks!

Is there a way to suppress analyzer warnings?   

```
dartanalyzer lib/d3.dart
Analyzing [lib/d3.dart]...
[warning] Missing concrete implementation of getter 'List.reversed', 'List.indexOf', 'List.removeRange', 'Iterable.contains' and 48 more (/Users/john/code/projects/d3dart/lib/d3.dart, line 11, col 7)
1 warning found.
```

**Comment by jacob314** [comment]
There isn't a way to suppress analyzer warnings. 
Its annoying and a little strange as the core List method implementations will take priority anyway but you could provide stub external implementations each dart List method to suppress the analyzer warnings.

I'm adding a fix so that writing
@JS()
class Foo extends List {
...
}
is allowed in Dart2Js. That version will work without analyzer warnings.

**Comment by johnpryan** [comment]
great, thanks!

**Comment by jacob314** [comment]
Actually... I've thought about this more and the cleanest solution for 1.13 is just to mark the class as abstract which make the Analyzer happy and works great in Dartium and Dart2js today.
The only thing you lose by having the class be abstract is you can't specify a constructor but a constructor makes little sense for an Array subclass in JS.
''
'@JS()
abstract class Foo implements List {
  // ...
 
}'''

**Comment by zoechi** [comment]
Maybe adding the `@proxy` annotation or implementing `noSuchMethod` (didn't have a closer look at the code)

**Comment by johnpryan** [comment]
`abstract class Selection implements List ` or `noSuchMethod` both get the job done.  Thanks!

**Comment by zoechi** [comment]
Ups, somehow missed the comment about `abstract`.

**Comment by jacob314** [comment]
JS interop classes cannot implement noSuchMethod. Maybe adding @proxy would work. I haven't tried that.

**Comment by a14n** [comment]
Couldn't we use `external factory` to avoid warning?

IIUC `external factory` means anonymous objects regardless `@JS` has a name specified or not. We could change this behaviour to make `external factory Foo();` like `external Foo();` if `@JS` is use with a name.

Thus the above example would be :

```dart
@JS('Foo')
abstract class Foo implements List {
  external factory Foo();
}
```

What do you think?

**Comment by jacob314** [comment]
@a14n that is a good point. Currently I've been using/abusing "external factory" as constructors that desugar to object literals but that is somewhat magical and prevents this use of external factory to avoid spurious Analyzer warnings. Any thoughts on adding a '''@Literal''' annotation to package:js that indicates that a constructor should be implemented by a JavaScript object literal? Then we could just support using "external factory" as the default pattern for general purpose JavaScript constructors.

**Comment by a14n** [comment]
We could also have another constructor like `@JS.literal()`.

**Comment by jacob314** [comment]
Added
@anonymous
specified at the class level.
Landed for 1.13

On Thu, Oct 29, 2015 at 7:01 AM, Alexandre Ardhuin <notifications@github.com
